statsfs/otelstats/subsys_metric_parser.go

```diff
@@ -0,0 +1,158 @@
+package otelstats
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// MetricInfo contains a Label used to identify the specific device
+// and a Path from where the metric for this device could be retrieved
+type MetricInfo struct {
+	Name  string
+	Label string
+	Path  string
+}
+
+// SubsysMetrics is a struct that represents metrics of a subsystem.
+// Path gives the base path to the subsystem's stats in statsfs
+// (usually /sys/kernel/stats/subsystemName), Metrics is a map with key
+// being the metric name, and value being a list of labels (devices with
+// the metric registered)
+type SubsysMetrics struct {
+	StatsfsPath   string
+	SubSystemName string
+	SubSystemPath string
+	Metrics       map[string][]MetricInfo
+}
+
+// newSubsysMetric creates a SubsysMetric struct given the mounting
+// point of statsfs filesystem (statsfsPath) and the subsystemName
+func newSubsysMetric(statsfsPath, subsystemName string) SubsysMetrics {
+	subsysMetrics := SubsysMetrics{
+		StatsfsPath:   statsfsPath,
+		SubSystemName: subsystemName,
+		SubSystemPath: strings.Join([]string{statsfsPath, subsystemName}, "/"),
+		Metrics:       make(map[string][]MetricInfo),
+	}
+	subsysMetrics.constructMetricMap()
+	return subsysMetrics
+}
+
+func (m *SubsysMetrics) constructMetricMap() {
+	err := filepath.Walk(m.SubSystemPath, m.updateMetricMap)
+	if err != nil {
+		log.Printf("Failed to parse metrics for subsystem %v at %v", m.SubSystemName, m.SubSystemPath)
+	}
+}
+
+func (m *SubsysMetrics) updateMetricMap(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		log.Printf("Failed to walk to file %v\n", path)
```

Comment:
I somehow missed this in the last review. Let's propagate the error to `newSubsysMetric` instead of simply logging them here. And then either log or return the error at line 143